docs/docs/dev_guide/validation_plugins.md

```diff
@@ -8,56 +8,49 @@ has_children: false
-     ancestor_definitions = get_definition_ancestry(definition, language_context) 
-     sub_structure_definitions = get_definition_schema_components(definition, language_context) 
-     all_applicable_definitions = ancestor_definitions + sub_structure_definitions 
-  
-     for target_schema_definition in all_applicable_definitions: 
-         sub_definition_validations = target_schema_definition.get_validations() 
-  
-         if sub_definition_validations: 
-             for validation in sub_definition_validations: 
-                 validation_name = validation.get(DEFINITION_FIELD_NAME) 
-                 validator_plugin = list(filter(lambda plugin: plugin.name == validation_name, applicable_validator_plugins)) 
-  
-                 if validator_plugin: 
-                     validator_results.append( 
-                         _apply_validator(definition, target_schema_definition, language_context, validator_plugin[0]) 
-                     ) 
-  
-     validator_results.append(_validate_primitive_types(definition, language_context)) 
-     return validator_results 
+ def _validate_definition(
+     definition: Definition, validator_plugins: list[DefinitionValidationContribution], language_context: LanguageContext
+ ) -> list[ValidatorResult]:
+     """Traverse the definition and validate it according to the validator plugins."""
+
+     validator_results = []
+
+     # Decompose the definition under validation into any other definitions that compose its sub-structures
+     applicable_validator_plugins = get_applicable_validators_for_definition(definition, validator_plugins, language_context)
+     ancestor_definitions = get_definition_ancestry(definition, language_context)
+     sub_structure_definitions = get_definition_schema_components(definition, language_context)
+     all_applicable_definitions = ancestor_definitions + sub_structure_definitions
+
+     # For every identified sub-structure that is defined by another schema definition, validate it per its corresponding validations
+     for target_schema_definition in all_applicable_definitions:
+         sub_definition_validations = target_schema_definition.get_validations()
+
+         if sub_definition_validations:
+             for validation in sub_definition_validations:
+                 validation_name = validation.get(DEFINITION_FIELD_NAME)
+                 validator_plugin = list(filter(lambda plugin: plugin.name == validation_name, applicable_validator_plugins))
+
+                 if validator_plugin:
+                     # Add the validation results to the overall collection of validation results
+                     validator_results.append(
+                         _apply_validator(definition, target_schema_definition, language_context, validator_plugin[0])
+                     )
+
+     validator_results.append(_validate_primitive_types(definition, language_context))
+     return validator_results
```

Comment:
I think we should just link to this as it'd be too easy for this snippet to get out of sync with the actual implementation.

Reply:
Captured in #592